atc/scheduler/scheduler.go

```diff
@@ -93,9 +93,34 @@ func (s *Scheduler) ensurePendingBuildExists(
 		}
 	}
 
+	jobNeedsManualTrigger := firstOccurrenceInInputMapping(inputMapping) && jobIsManuallyTriggered(job)
+	if job.NeedsManualTrigger() != jobNeedsManualTrigger {
+		if err := job.SetNeedsManualTrigger(jobNeedsManualTrigger); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+func firstOccurrenceInInputMapping(inputMapping algorithm.InputMapping) bool {
```

Comment:
I think this could be removed with a slight refactor of the loop that goes over the inputs:

```go
var needsTrigger bool
for _, inputConfig := range job.Config().Inputs() {
	inputVersion, ok := inputMapping[inputConfig.Name]
	if !ok {
		continue
	}

	if !inputVersion.FirstOccurrence {
		continue
	}

	needsTrigger = true

	if inputConfig.Trigger {
		needsTrigger = false

		// ensure pending build exists

		break
	}
}

if job.NeedsManualTrigger() != needsTrigger {
	// set
}
```

Reply:
We tried this approach but in the end we both felt the readability suffered.